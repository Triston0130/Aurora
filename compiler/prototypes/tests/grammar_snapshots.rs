use insta::assert_snapshot;
use std::fs;

fn read_example(name: &str) -> String {
    fs::read_to_string(format!("examples/{}.aur", name)).expect("example file")
}

#[test]
fn parse_actor_pipeline() {
    let source = read_example("actor_pipeline");
    let ast = aurora_compiler_prototypes::parser::parse_module(&source).expect("parse");
    assert_snapshot!(format!("{ast:#?}"));
}

#[test]
fn parse_gpu_zone() {
    let source = read_example("gpu_zone");
    let ast = aurora_compiler_prototypes::parser::parse_module(&source).expect("parse");
    assert_snapshot!(format!("{ast:#?}"));
}

#[test]
fn parse_effectful_trait() {
    let source = read_example("effectful_trait");
    let ast = aurora_compiler_prototypes::parser::parse_module(&source).expect("parse");
    assert_snapshot!(format!("{ast:#?}"));
}

#[test]
fn parse_zone_decl() {
    let source = read_example("zone_decl");
    let ast = aurora_compiler_prototypes::parser::parse_module(&source).expect("parse");
    assert_snapshot!(format!("{ast:#?}"));
}

#[test]
fn parse_patterns() {
    let source = read_example("patterns");
    let ast = aurora_compiler_prototypes::parser::parse_module(&source).expect("parse");
    assert_snapshot!(format!("{ast:#?}"));
}

#[test]
fn parse_types_features() {
    let source = read_example("types_features");
    let ast = aurora_compiler_prototypes::parser::parse_module(&source).expect("parse");
    assert_snapshot!(format!("{ast:#?}"));
}

#[test]
fn parse_use_groups() {
    let source = read_example("use_groups");
    let ast = aurora_compiler_prototypes::parser::parse_module(&source).expect("parse");
    assert_snapshot!(format!("{ast:#?}"));
}

#[test]
fn parse_effect_handler() {
    let source = read_example("effect_handler");
    let ast = aurora_compiler_prototypes::parser::parse_module(&source).expect("parse");
    assert_snapshot!(format!("{ast:#?}"));
}
