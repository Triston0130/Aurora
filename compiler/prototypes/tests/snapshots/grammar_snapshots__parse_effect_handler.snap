---
source: compiler/prototypes/tests/grammar_snapshots.rs
expression: "format!(\"{ast:#?}\")"
---
Module {
    id: NodeId(
        0,
    ),
    span: Span {
        file: FileId(
            0,
        ),
        start: Position {
            offset: 0,
            line: 0,
            column: 0,
        },
        end: Position {
            offset: 0,
            line: 0,
            column: 0,
        },
        line: 0,
        column: 0,
    },
    docs: [],
    items: [
        Function(
            FunctionDecl {
                attributes: [],
                docs: [],
                visibility: Private,
                signature: FunctionSig {
                    constness: false,
                    asyncness: false,
                    name: "main",
                    generics: GenericParams {
                        params: [],
                    },
                    params: [],
                    return_type: None,
                    effect: None,
                    where_clause: [],
                },
                body: Block {
                    statements: [],
                    tail: Some(
                        Expr {
                            kind: Handle {
                                body: Expr {
                                    kind: Block(
                                        Block {
                                            statements: [],
                                            tail: Some(
                                                Expr {
                                                    kind: Async(
                                                        Block {
                                                            statements: [],
                                                            tail: Some(
                                                                Expr {
                                                                    kind: Literal(
                                                                        Unit,
                                                                    ),
                                                                    span: Span {
                                                                        file: FileId(
                                                                            0,
                                                                        ),
                                                                        start: Position {
                                                                            offset: 53,
                                                                            line: 4,
                                                                            column: 13,
                                                                        },
                                                                        end: Position {
                                                                            offset: 55,
                                                                            line: 4,
                                                                            column: 15,
                                                                        },
                                                                        line: 4,
                                                                        column: 13,
                                                                    },
                                                                },
                                                            ),
                                                            span: Span {
                                                                file: FileId(
                                                                    0,
                                                                ),
                                                                start: Position {
                                                                    offset: 39,
                                                                    line: 3,
                                                                    column: 15,
                                                                },
                                                                end: Position {
                                                                    offset: 65,
                                                                    line: 5,
                                                                    column: 10,
                                                                },
                                                                line: 3,
                                                                column: 15,
                                                            },
                                                        },
                                                    ),
                                                    span: Span {
                                                        file: FileId(
                                                            0,
                                                        ),
                                                        start: Position {
                                                            offset: 33,
                                                            line: 3,
                                                            column: 9,
                                                        },
                                                        end: Position {
                                                            offset: 65,
                                                            line: 5,
                                                            column: 10,
                                                        },
                                                        line: 3,
                                                        column: 9,
                                                    },
                                                },
                                            ),
                                            span: Span {
                                                file: FileId(
                                                    0,
                                                ),
                                                start: Position {
                                                    offset: 23,
                                                    line: 2,
                                                    column: 12,
                                                },
                                                end: Position {
                                                    offset: 71,
                                                    line: 6,
                                                    column: 6,
                                                },
                                                line: 2,
                                                column: 12,
                                            },
                                        },
                                    ),
                                    span: Span {
                                        file: FileId(
                                            0,
                                        ),
                                        start: Position {
                                            offset: 0,
                                            line: 0,
                                            column: 0,
                                        },
                                        end: Position {
                                            offset: 0,
                                            line: 0,
                                            column: 0,
                                        },
                                        line: 0,
                                        column: 0,
                                    },
                                },
                                handlers: [
                                    EffectHandler {
                                        label: "Async",
                                        body: Block {
                                            statements: [],
                                            tail: Some(
                                                Expr {
                                                    kind: Literal(
                                                        Unit,
                                                    ),
                                                    span: Span {
                                                        file: FileId(
                                                            0,
                                                        ),
                                                        start: Position {
                                                            offset: 110,
                                                            line: 8,
                                                            column: 13,
                                                        },
                                                        end: Position {
                                                            offset: 112,
                                                            line: 8,
                                                            column: 15,
                                                        },
                                                        line: 8,
                                                        column: 13,
                                                    },
                                                },
                                            ),
                                            span: Span {
                                                file: FileId(
                                                    0,
                                                ),
                                                start: Position {
                                                    offset: 96,
                                                    line: 7,
                                                    column: 18,
                                                },
                                                end: Position {
                                                    offset: 122,
                                                    line: 9,
                                                    column: 10,
                                                },
                                                line: 7,
                                                column: 18,
                                            },
                                        },
                                        span: Span {
                                            file: FileId(
                                                0,
                                            ),
                                            start: Position {
                                                offset: 87,
                                                line: 7,
                                                column: 9,
                                            },
                                            end: Position {
                                                offset: 92,
                                                line: 7,
                                                column: 14,
                                            },
                                            line: 7,
                                            column: 9,
                                        },
                                    },
                                ],
                            },
                            span: Span {
                                file: FileId(
                                    0,
                                ),
                                start: Position {
                                    offset: 16,
                                    line: 2,
                                    column: 5,
                                },
                                end: Position {
                                    offset: 128,
                                    line: 10,
                                    column: 6,
                                },
                                line: 2,
                                column: 5,
                            },
                        },
                    ),
                    span: Span {
                        file: FileId(
                            0,
                        ),
                        start: Position {
                            offset: 10,
                            line: 1,
                            column: 11,
                        },
                        end: Position {
                            offset: 130,
                            line: 11,
                            column: 2,
                        },
                        line: 1,
                        column: 11,
                    },
                },
                span: Span {
                    file: FileId(
                        0,
                    ),
                    start: Position {
                        offset: 0,
                        line: 1,
                        column: 1,
                    },
                    end: Position {
                        offset: 2,
                        line: 1,
                        column: 3,
                    },
                    line: 1,
                    column: 1,
                },
                id: NodeId(
                    1,
                ),
            },
        ),
    ],
}
